Source/EasyNetQ/Internals/DisposableAction.cs

```diff
@@ -0,0 +1,26 @@
+using System;
+
+namespace EasyNetQ.Internals
+{
+    internal struct DisposableAction<TState> : IDisposable
+    {
+        public static DisposableAction<TState> Create(Action<TState> action, TState state)
+        {
+            return new DisposableAction<TState>(action, state);
+        }
+
+        private readonly Action<TState> action;
+        private readonly TState state;
+
+        private DisposableAction(Action<TState> action, TState state)
```

Comment:
Why not public ? Why do `Create` method?

Reply:
Because I don't want to keep this struct forever(I mean to next major version). Create is needed for type inference.

Reply:
OK. I asked about public ctor, not whole class.

> Create is needed for type inference.

Type inference does not work in this case:
```c#
DisposableAction<Guid>.Create(DeRegisterResponseActions, correlationId);
```
Usually one have to write a helper non-generic class with generic method inside to make type inference work.

Reply:
You are right. I have uncommitted this one:
```
    internal static class DisposableActions
    {
        public static DisposableAction<TState> Create<TState>(Action<TState> action, TState state)
        {
            return new DisposableAction<TState>(action, state);
        }
    }

    internal readonly struct DisposableAction<TState> : IDisposable
    {
        private readonly Action<TState> action;
        private readonly TState state;

        public DisposableAction(Action<TState> action, TState state)
        {
            this.action = action;
            this.state = state;
        }

        public void Dispose()
        {
            action(state);
        }
    }
```